### ISSUE TITLE
Cannot get custom rules to work

### ISSUE THREAD

**Issue by smeymannlt** [detekt/detekt#3897]
## Expected Behavior
With the custom rules as below, it should catch a "CodeSmell" on any appcompat import.

## Observed Behavior
- No effect to CustomRulesProvider. 
- No error when spoiling META-INF/../RuleSetProvider
- No `projects` in the gradle
- No plugin `module`
- No rule is available to define at the config

## Steps to Reproduce
Gradle

```
plugins {
    id("com.android.application")
    id("kotlin-android")
    id("io.gitlab.arturbosch.detekt").version("1.17.1")
}

android {
    compileSdkVersion (30)
    buildToolsVersion ="30.0.3"

    defaultConfig {
        applicationId = "com.playgrounds.detektplayground"
        minSdkVersion (26)
        targetSdkVersion (30)
        versionCode = 1
        versionName = "1.0"

        testInstrumentationRunner = "androidx.test.runner.AndroidJUnitRunner"
    }

    compileOptions {
        sourceCompatibility=JavaVersion.VERSION_1_8
        targetCompatibility=JavaVersion.VERSION_1_8
    }
    kotlinOptions {
        jvmTarget = "1.8"
    }
}

detekt {
    toolVersion = Versions.detekt
    input = files("src/main/java")
    config = files("../detekt/detekt-config.yml")
    reports {
        txt {
            enabled = true
            destination = file("detekt.txt")
        }
        html {
            enabled = true
            destination = file("detekt.html")
        }
    }
}

dependencies {
    val detektVersion = Versions.detekt // 1.17.1
    detektPlugins("io.gitlab.arturbosch.detekt:detekt-formatting:$detektVersion")
    compileOnly("io.gitlab.arturbosch.detekt:detekt-api:$detektVersion")
    detekt("io.gitlab.arturbosch.detekt:detekt-cli:$detektVersion")



    implementation("org.jetbrains.kotlin:kotlin-stdlib:${Versions.kotlin}")
    implementation("androidx.core:core-ktx:1.5.0")
    implementation("androidx.appcompat:appcompat:1.3.0")
    implementation("com.google.android.material:material:1.3.0")
    implementation("androidx.constraintlayout:constraintlayout:2.0.4")
    testImplementation("junit:junit:4.+")
    androidTestImplementation("androidx.test.ext:junit:1.1.2")
    androidTestImplementation("androidx.test.espresso:espresso-core:3.3.0")
    testImplementation("io.gitlab.arturbosch.detekt:detekt-api:$detektVersion")
    testImplementation("io.gitlab.arturbosch.detekt:detekt-test:$detektVersion")
}
```

The rule
```
package com.playgrounds.detekt.customrule

import io.gitlab.arturbosch.detekt.api.*
import io.gitlab.arturbosch.detekt.api.internal.ActiveByDefault
import org.jetbrains.kotlin.psi.KtImportDirective

@ActiveByDefault(since = "now") // the value for "since" is not documented.
class SomeCustomRule(config: Config) : Rule(config) {
    override val issue: Issue
        get() = Issue("ImportThingy", Severity.CodeSmell, "I don't like this string", Debt(10))

    override fun visitImportDirective(importDirective: KtImportDirective) {
        val import: String = importDirective.importPath?.pathStr ?: ""
        println("Import $import")
        if ("appcompat" in import) {
            report(CodeSmell(
                issue,
                Entity.from(importDirective),
                "Importing $import which is an internal import"))
        }
    }
}
```

The provider
```
package com.playgrounds.detekt.customrule

import io.gitlab.arturbosch.detekt.api.Config
import io.gitlab.arturbosch.detekt.api.RuleSet
import io.gitlab.arturbosch.detekt.api.RuleSetProvider

class CustomRuleSetProvider : RuleSetProvider {
    override val ruleSetId: String = "detekt-custom-rules"

    override fun instance(config: Config) = RuleSet(ruleSetId, listOf(SomeCustomRule(config)))
}

```

And, last is the pointer
```
src/main/resources/META-INF/services/io.gitlab.arturbosch.detekt.api.RuleSetProvider
```

that contains the path to CustomRuleSetProvider
```
com.playgrounds.detekt.customrule.CustomRuleSetProvider
```

[Attached is the project Detektplayground.zip](https://github.com/detekt/detekt/files/6710694/Detektplayground.zip)

## Context
I am planning to use Detekt, and create my own rules.

## Your Environment
* Version of detekt used: 1.17.1
* Version of Gradle used: 4.2.1
* Gradle scan link (add `--scan` option when running the gradle task): https://gradle.com/s/kbghwsodarhg4
* Operating System and version: MacOs 11.4.
* Link to your project (if it's a public repository):

Sorry if it's a silly thing I missed. I have passed through the instructions a few times and - it just won't work.

**Comment by cortinico** [comment]
> * No effect to CustomRulesProvider.
> * No error when spoiling META-INF/../RuleSetProvider
> * No `projects` in the gradle
> * No plugin `module`
> * No rule is available to define at the config

I'm not sure I understand what you mean here.
Can you please elaborate?

I invite you to read this page https://detekt.github.io/detekt/extensions.html where we explain how to write custom rules.

**Comment by smeymannlt** [comment]
Hi, Thanks.... 

> I invite you to read this page https://detekt.github.io/detekt/extensions.html where we explain how to write custom rules.

It's the manual, of course. That's what I followed in the first place. 

The thing is that I created a silly test rule that would exclaim whenever an appcompat is found in one of the imports.
```if ("appcompat" in import) { report(CodeSmell(...```
I followed the manual as can be seen from the snippets.
But when running `gradlew detekt` nothing happens. In order to debug, I tried to check if Detekt finds my rule. 

- I tried to deliberately "destroy" the provider adding gibberish to it. No error
- I tried to rename CustomRuleSetProvider to Custo_mRuleSetProvider to make the file in META-INF contain a wrong pointer. Nothing. No failures. Conclusion: My rule provider is not started.
- I checked other manuals and the example code, and tried to add those `detektPlugins project(":detekt-custom-rules")` lines to Gradle. Of course Gradle fails.

Seems like it's some major (and probably elementary) thing that I forgot, or missing at the manuals.
Still I don't understand why it just won't work. In any way.

Thanks, and sorry if it's a false positive bug.

**Comment by cortinico** [comment]
>     * I checked other manuals and the example code, and tried to add those `detektPlugins project(":detekt-custom-rules")` lines to Gradle. Of course Gradle fails.

So that's the culprit. What's the failure?

Detekt works that finds all the JARs imported with `detektPlugins` for external rules. If you don't add a dependency to your project the ruleset won't be considered.

**Comment by smeymannlt** [comment]
Okay, so the failure is that my rule and provider are not applied (or even parsed). I even changed the rule to:

```
    override fun visitImportDirective(importDirective: KtImportDirective) {
        val import: String = importDirective.importPath?.pathStr ?: ""
        report(CodeSmell(
            issue,
            Entity.from(importDirective),
            "Importing $import which is an internal import"))
    }
```

But it doesn't report. It doesn't do anything. 
The failure is that I don't see it fail. I expect it to report a code-smell, but it won't.

It's because I can't make it find my rule.

I can't make detektPlugins compile in Gradle.
<img width="733" alt="image" src="https://user-images.githubusercontent.com/69756356/123476880-f3e53f00-d605-11eb-8bdd-12b5c39481c3.png">


(sorry, many imports, was desperate 😞) 

If I add it, I get 

```
org.gradle.internal.exceptions.LocationAwareException: Build file '.../Detektplayground/app/build.gradle.kts' line: 59
Project with path ':detekt-custom-rules' could not be found in project ':app'.
	at org.gradle.kotlin.dsl.execution.InterpreterKt$locationAwareExceptionFor$2.invoke(Interpreter.kt:563)
	at org.gradle.kotlin.dsl.execution.InterpreterKt.locationAwareExceptionFor(Interpreter.kt:570)
	at org.gradle.kotlin.dsl.execution.InterpreterKt.locationAwareExceptionHandlingFor(Interpreter.kt:536)
	at org.gradle.kotlin.dsl.execution.InterpreterKt.access$locationAwareExceptionHandlingFor(Interpreter.kt:1)
	at org.gradle.kotlin.dsl.execution.Interpreter$ProgramHost.handleScriptException(Interpreter.kt:373)

```

Though my provider is
```
class CustomRuleSetProvider : RuleSetProvider {
    override val ruleSetId: String = "detekt-custom-rules"

    override fun instance(config: Config) = RuleSet(ruleSetId, listOf(SomeCustomRule(config)))
}
```

I think I am missing something big about custom rules. I understood that it's rules that you can define to do some extra checks on your code when you run `./gradelw detekt`. Am I wrong?


Thanks

**Comment by smeymannlt** [comment]
Let me try to explain it simply (I hope it'll come out simple 🙏 ). 

Created a dummy project at 
https://github.com/smeymannlt/DetektPlayground

There are a few failures here may be one simple thing I forgot:

1. Expected result: If all works - then this should always report a code-smell.
```
    override fun visitImportDirective(importDirective: KtImportDirective) {
        report(CodeSmell(issue, Entity.from(importDirective), "An import"))
    }
```
It doesn't.

2. If the provider is found by Detekt then - this should not compile.
```
class CustomRuleSetProvider : RuleSetProvider {
    GIBBERISH-DO-NOT-COMPILE
```
It always does.

3. If the plugin runs and creates my project - this should work when running `detekt` task.
```
detektPlugins(project(":detekt-custom-rules"))
```
 It won't.

4. And lastly, if Detekt reads the META-INF/services/io...RulesSetProvider file
```
src/main/resources/META-INF/services/io.gitlab.arturbosch.detekt.api.RuleSetProvider
```
Then if I modify it to contain gibberish such as
```
Snow.white.breaks.guitars.on.Jimmy.Hendrix
```
Nothing should compile. It does.

**Comment by cortinico** [comment]
I believe I missed a comment here @smeymannlt 

Please let us know if you solved so we can close the issue or not 👍

**Comment by smeymannlt** [comment]
Hi. 
> I believe I missed a comment here @smeymannlt

Yes, but there was an update 😊 

**Update:** After converting to build.gradle.kts and applying this line
```
tasks.withType<io.gitlab.arturbosch.detekt.Detekt> {dependsOn (":app:assemble")}
```
It now reads the rule. At least it parses the rule file. Doesn't enable it yet, even though it has 
```
@ActiveByDefault(since = "0.0.0")
class SomeCustomRule(config: Config) : Rule(config) {
```

Uploaded the playground at https://github.com/smeymannlt/DetektPlayground

Still, 
1. How do I write it on Groovy? `detektPlugins project(":app")` will fail.
2. Also building with no `:assemble` - will fail. `detektPlugins(project(":app"))` will fail, too.
3. It now reads the rule and the provider but if my provider has

```
override val ruleSetId: String = "detektcustomrules"
```
I can't enable the rule at my `detekt-config.yml`
```
detektcustomrules:
  active: true
  SomeRule:
    active: true
```

Because
```
Execution failed for task ':app:detekt'.
> [31mRun failed with 1 invalid config property.[0m
  	- [31mProperty 'detektcustomrules' is misspelled or does not exist.[0m

```
What am I missing?

Thanks

**Comment by cortinico** [comment]
> Uploaded the playground at [smeymannlt/DetektPlayground](https://github.com/smeymannlt/DetektPlayground)

Thanks for sharing this.

The current setup won't work as it is right now. Specifically you have to:
- Create a new JVM Gradle module and call it `detekt-custom-rules`
- Move all the custom rules code (the rule, the ruleset provider and the META-INF folder there)
- Un-comment this line: https://github.com/smeymannlt/DetektPlayground/blob/5e2ca2130d64146be958b97d56d36976cf5cf168/app/build.gradle.kts#L59
- Remove all the code you mentioned that relies on the `:assemble` dependencies, that is not needed.

The problem is that right now you have custom rules AND app code together in the same module. Detekt can't use a custom rule from the same module where you want to apply the inspection.

**Comment by smeymannlt** [comment]
Hi. Thanks
Sorry for the delayed answer... 
Tried to do that. 

`CustomRuleSetProvider` is compiled (if I put gibberish it will not compile), but never instantiated.

Because nothing happens when 
```
class CustomRuleSetProvider : RuleSetProvider {

    init {
        throw RuntimeException("INIT")
    }

```
Pushed an updated code.
It fails on the configuration file because the configuration yml contains `detekt-custom-rules` which are not known.
They are not known because class `CustomRuleSetProvider : RuleSetProvider` is not executed

**Comment by cortinico** [comment]
There are still a couple of things that will not make the setup work as it is:

https://github.com/smeymannlt/DetektPlayground/blob/f4d346e321fcdbbdc652b65369d417dd19940f45/detekt-custom-rules/build.gradle.kts#L2-L3

Rules should be in a Java/Kotlin library, and not an Android library.
So you need to replace the `plugins` block with:

```
plugins {
    kotlin("jvm")
    id("io.gitlab.arturbosch.detekt").version("1.17.1")
}
```

You should also remove the whole `android{}` block below.
As well as all the dependencies here: https://github.com/smeymannlt/DetektPlayground/blob/f4d346e321fcdbbdc652b65369d417dd19940f45/detekt-custom-rules/build.gradle.kts#L60-L65

Finally [this line](https://github.com/smeymannlt/DetektPlayground/blob/f4d346e321fcdbbdc652b65369d417dd19940f45/app/build.gradle.kts#L59) should just be:

```
detektPlugins(project(":detekt-custom-rules"))
```

**Comment by smeymannlt** [comment]
Ouch 🤦 Silly me! Thanks a lot! It works now. 

You guys rock 🎸 !

I think I am going to prepare a "guide for beginners" that would summarize it all as a cookbook. Will write it in md and contribute to you fellas.

Thanks a lot

**Comment by cortinico** [comment]
> I think I am going to prepare a "guide for beginners" that would summarize it all as a cookbook. Will write it in md and contribute to you fellas.

Great this will be more than appreciated 🙏
